Merge basic blocks where possible when generating LLVM IR.

nnethercote opened the pull request:
r? @ghost

nnethercote committed:
Merge basic blocks where possible when generating LLVM IR.

In `codegen_assert_terminator` we decide if a BB's successor is a
candidate for merging, which requires that it be the only successor, and
that it only have one predecessor. That result then gets passed down,
and if it reaches `funclet_br` with the appropriate BB characteristics,
then no `br` instruction is issued, a `MergingSucc::True` result is
passed back, and the merging proceeds in `codegen_block`.

The commit also adds `CachedLlbb`, a new type to help keep track of
each BB that has been merged into its predecessor.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit a48b4cc08278a9a37892cc745a38b1cbfbf29340 with merge 0a35b2797788a7dd1063c4b0155bc4ade8ec24f5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a48b4cc08278a9a37892cc745a38b1cbfbf29340","merge_sha":"0a35b2797788a7dd1063c4b0155bc4ade8ec24f5"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3263100037/jobs/5361310540) [(plain)](https://github.com/rust-lang/rust/commit/a48b4cc08278a9a37892cc745a38b1cbfbf29340/checks/8926898986/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/block.rs at line 131:
         fx: &mut FunctionCx<'a, 'tcx, Bx>,
         bx: &mut Bx,
-        follow_on: bool
+        follow_on: bool,
     ) -> bool {
     ) -> bool {
         // njn: duplicated stuff from .lltarget()
         // njn: also, some non-(None,None) cases in .lltarget() that could be used here
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/block.rs at line 834:
         if intrinsic == Some(sym::caller_location) {
             if let Some(target) = target {
-                let location = self
-                let location = self
-                    .get_caller_location(bx, mir::SourceInfo { span: fn_span, ..source_info });
+                let location =
+                    self.get_caller_location(bx, mir::SourceInfo { span: fn_span, ..source_info });
 
                 if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
                     location.val.store(bx, tmp);
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/block.rs at line 1019:
             self.codegen_argument(bx, op, &mut llargs, &fn_abi.args[i]);
         }
         let num_untupled = untuple.map(|tup| {
-            self.codegen_arguments_untupled(
-                bx,
-                tup,
-                &mut llargs,
-                &fn_abi.args[first_args.len()..],
-            )
+            self.codegen_arguments_untupled(bx, tup, &mut llargs, &fn_abi.args[first_args.len()..])
 
         let needs_location =
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/block.rs at line 1330:
                     target,
                     target,
                     cleanup,
                     fn_span,
-                    follow_on
+                    follow_on,
                 )
             }
             mir::TerminatorKind::GeneratorDrop | mir::TerminatorKind::Yield { .. } => {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/mod.rs" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/_impl.rs" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/arg_matrix.rs" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/checks.rs" "/checkout/compiler/rustc_hir_analysis/src/check/method/prelude2021.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/block.rs" "/checkout/compiler/rustc_codegen_ssa/src/base.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3263102982/jobs/5364078209)
Build commit: 0a35b2797788a7dd1063c4b0155bc4ade8ec24f5 (`0a35b2797788a7dd1063c4b0155bc4ade8ec24f5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3263102982/jobs/5364078209"},"merge_sha":"0a35b2797788a7dd1063c4b0155bc4ade8ec24f5"} -->

rust-timer commented:
Queued 0a35b2797788a7dd1063c4b0155bc4ade8ec24f5 with parent 1536ab1b383f21b38f8d49230a2aecc51daffa3d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1536ab1b383f21b38f8d49230a2aecc51daffa3d&end=0a35b2797788a7dd1063c4b0155bc4ade8ec24f5).

rust-timer commented:
Finished benchmarking commit (0a35b2797788a7dd1063c4b0155bc4ade8ec24f5): [comparison URL](https://perf.rust-lang.org/compare.html?start=1536ab1b383f21b38f8d49230a2aecc51daffa3d&end=0a35b2797788a7dd1063c4b0155bc4ade8ec24f5&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1536ab1b383f21b38f8d49230a2aecc51daffa3d&end=0a35b2797788a7dd1063c4b0155bc4ade8ec24f5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 1.0%     | [0.8%, 1.2%] | 6         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -0.3%    | [-0.3%, -0.3%] | 1         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 0.8%     | [-0.3%, 1.2%] | 7         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1536ab1b383f21b38f8d49230a2aecc51daffa3d&end=0a35b2797788a7dd1063c4b0155bc4ade8ec24f5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 9.2%     | [9.2%, 9.2%] | 1         |
| Improvements ✅ <br /> (primary) | -0.1%    | [-0.1%, -0.1%] | 1         |
| Improvements ✅ <br /> (secondary) | -2.5%    | [-3.2%, -2.1%] | 4         |
| All ❌✅ (primary) | -0.1%    | [-0.1%, -0.1%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1536ab1b383f21b38f8d49230a2aecc51daffa3d&end=0a35b2797788a7dd1063c4b0155bc4ade8ec24f5&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

nnethercote commented:
The instruction count results aren't a win, but there are hints of goodness in the results for cycles, wall-time, max-rss, and especially binary size. The current version only merges the simplest cases, and there are quite a few more cases that can be handled, so I will continue working on them.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 165b498be31961a522cedd64bb9bbe33c072d0f4 with merge 61e75799adaa22db3b3d115e5c1d921210da60ad...
<!-- homu: {"type":"TryBuildStarted","head_sha":"165b498be31961a522cedd64bb9bbe33c072d0f4","merge_sha":"61e75799adaa22db3b3d115e5c1d921210da60ad"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3271108651/jobs/5382500064)
Build commit: 61e75799adaa22db3b3d115e5c1d921210da60ad (`61e75799adaa22db3b3d115e5c1d921210da60ad`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3271108651/jobs/5382500064"},"merge_sha":"61e75799adaa22db3b3d115e5c1d921210da60ad"} -->

rust-timer commented:
Queued 61e75799adaa22db3b3d115e5c1d921210da60ad with parent 98a5ac269cffada469753ad2416717e251863f9a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=98a5ac269cffada469753ad2416717e251863f9a&end=61e75799adaa22db3b3d115e5c1d921210da60ad).

rust-timer commented:
Finished benchmarking commit (61e75799adaa22db3b3d115e5c1d921210da60ad): [comparison URL](https://perf.rust-lang.org/compare.html?start=98a5ac269cffada469753ad2416717e251863f9a&end=61e75799adaa22db3b3d115e5c1d921210da60ad&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=98a5ac269cffada469753ad2416717e251863f9a&end=61e75799adaa22db3b3d115e5c1d921210da60ad&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.9%     | [0.4%, 1.3%] | 7         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -0.3%    | [-0.3%, -0.3%] | 2         |
| Improvements ✅ <br /> (secondary) | -0.3%    | [-0.3%, -0.3%] | 1         |
| All ❌✅ (primary) | 0.6%     | [-0.3%, 1.3%] | 9         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=98a5ac269cffada469753ad2416717e251863f9a&end=61e75799adaa22db3b3d115e5c1d921210da60ad&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 4.0%     | [2.2%, 6.3%] | 5         |
| Improvements ✅ <br /> (primary) | -1.5%    | [-2.7%, -0.3%] | 2         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -1.5%    | [-2.7%, -0.3%] | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=98a5ac269cffada469753ad2416717e251863f9a&end=61e75799adaa22db3b3d115e5c1d921210da60ad&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

nnethercote commented:
To summarize:
- MIR uses one definition of BBs, and LLVM IR uses another.
- rustc generates reasonable MIR code.
- rustc does a 1-to-1 translation of MIR BBs to LLVM IR BBs, which is reasonable.
- The resulting LLVM IR looks a bit silly and quite sub-optimal, because of the different BB definition.
- The sub-optimality doesn't end up mattering much in terms of compiler perf.
- The sub-optimality also doesn't matter for the output of opt builds, because LLVM can optimize away the extra jumps and the output ends up the same.
- The sub-optimality matters slightly for the output of debug builds, because it causes binaries to be about 0.5% bigger. It may also make them slightly slower, though I haven't measured that and I suspect the effect would be very small, probably less than 0.5%.

nnethercote committed:
Use `&mut Bx` more.

For the next commit, `FunctionCx::codegen_*_terminator` need to take a
`&mut Bx` instead of consuming a `Bx`. This triggers a cascade of
similar changes across multiple functions. The resulting code is more
concise and replaces many `&mut bx` expressions with `bx`.

nnethercote commented:
I'm reopening this for further consideration. Even though it didn't make much difference to compiler perf, which was the original motivation, it might still be worth merging.

Pros:
- Small (mostly <1%) binary size improvements, mostly for debug builds.
- Generated LLVM IR is easier to read, due to not having unnecessary branches.
- Generate binary is easier to debug at the machine code level, due to not having unnecessary branches (cc @Amanieu).

Cons:
- Some extra complexity in codegen.
- It's a little harder to see the MIR-to-LLVM-IR mapping.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9ca699a18569ad6117e593295e8da10d87c58db3 with merge cfe97392ed166ef3911f1539074b18cfe078baa5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9ca699a18569ad6117e593295e8da10d87c58db3","merge_sha":"cfe97392ed166ef3911f1539074b18cfe078baa5"} -->

nnethercote commented:
r? @bjorn3

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3425917808/jobs/5708659338)
Build commit: cfe97392ed166ef3911f1539074b18cfe078baa5 (`cfe97392ed166ef3911f1539074b18cfe078baa5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3425917808/jobs/5708659338"},"merge_sha":"cfe97392ed166ef3911f1539074b18cfe078baa5"} -->

rust-timer commented:
Queued cfe97392ed166ef3911f1539074b18cfe078baa5 with parent bc2504a83ca6ee8f6717dedd0721b90ffcbe1300, future [comparison URL](https://perf.rust-lang.org/compare.html?start=bc2504a83ca6ee8f6717dedd0721b90ffcbe1300&end=cfe97392ed166ef3911f1539074b18cfe078baa5).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (cfe97392ed166ef3911f1539074b18cfe078baa5): [comparison URL](https://perf.rust-lang.org/compare.html?start=bc2504a83ca6ee8f6717dedd0721b90ffcbe1300&end=cfe97392ed166ef3911f1539074b18cfe078baa5&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=bc2504a83ca6ee8f6717dedd0721b90ffcbe1300&end=cfe97392ed166ef3911f1539074b18cfe078baa5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.5%, -0.2%] | 10    |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.4%, -0.3%] | 2     |
| All ❌✅ (primary)                 | -0.7% | [-1.5%, -0.2%] | 10    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bc2504a83ca6ee8f6717dedd0721b90ffcbe1300&end=cfe97392ed166ef3911f1539074b18cfe078baa5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 5.5%  | [5.5%, 5.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.6%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -1.6% | [-2.6%, -0.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bc2504a83ca6ee8f6717dedd0721b90ffcbe1300&end=cfe97392ed166ef3911f1539074b18cfe078baa5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.0%, -3.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

bjorn3 commented on compiler/rustc_codegen_ssa/src/mir/mod.rs:
```diff
@@ -16,6 +16,18 @@ use rustc_middle::mir::traversal;
+    /// Has been created.
+    Some(T),
+
+    /// Nothing created yet, and nothing should be.
```
```suggestion
    /// Nothing should be created.
```

bjorn3 reviewed (commented):
How is it ensured that a basic block that will be merged in the future is not codegened at an earlier point? While it doesn't happen every time due to traversing mir in reverse postorder, it can still happen and I believe there are assumptions in cg_ssa about blocks being codegened only once somewhere in the code that avoids alloca where possible.

nnethercote commented:
> How is it ensured that a basic block that will be merged in the future is not codegened at an earlier point?

I think the reverse postorder traversal guarantees this? In combination with the constraint that merging is only possible for BB0 -> BB1 when BB1 is the only successor of BB0 and BB0 is the only predecessor of BB1.

I just added an assertion to check this -- when the successor is marked as `CachedLllbb::Skip`, the assertion checks that its current state is `CachedLllbb::None`. The full test suite passes with this assertion in place.

bjorn3 commented:
Right, makes sense.

@bors r+

bors commented:
:pushpin: Commit 1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79 -->
<!-- homu: {"type":"Approved","sha":"1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79 with merge d9fb02d529137dd32b71fa7b00176fc7a2c0ffa8...
<!-- homu: {"type":"BuildStarted","head_sha":"1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79","merge_sha":"d9fb02d529137dd32b71fa7b00176fc7a2c0ffa8"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3448930269/jobs/5756405400) [(plain)](https://github.com/rust-lang-ci/rust/commit/d9fb02d529137dd32b71fa7b00176fc7a2c0ffa8/checks/9445639972/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (80/84)
....       (84/84)


/checkout/src/test/rustdoc-gui/code-sidebar-toggle.goml code-sidebar-toggle... FAILED
[ERROR] (line 4) Error: Execution context was destroyed, most likely because of a navigation.: for command `wait-for: "#sidebar-toggle"`
Build completed unsuccessfully in 0:02:08
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3448930269/jobs/5756597313)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3448930269/jobs/5756597313","builder_name":"checks-actions"} -->

nnethercote commented:
No idea about that failure. Nothing obviously related to this PR's changes. Let's try again.

@bors retry rollup=never

nnethercote commented:
@bors retry

bors commented:
:hourglass: Testing commit 1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79 with merge f364a91e8c308b24eb748271fb6c12ced888f660...
<!-- homu: {"type":"BuildStarted","head_sha":"1a945aec45cb8f9c6dc0fce2bfd30a47fac2ac79","merge_sha":"f364a91e8c308b24eb748271fb6c12ced888f660"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3472336006/jobs/5803707081)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3472336006/jobs/5803707081","builder_name":"checks-actions"} -->

riking commented:
Failed test is debuginfo/lexical-scope-in-if-let.rs, test error is `error: line not found in debugger output: [...]z = 0n10` but that line is clearly in the debugger output:

```
0:000>  dv
              z = 0n10
              c = 0n789
```

nnethercote commented:
I have never seen these debuginfo tests before. But I notice that this test has four `cbd-command`/`cbd-check` groups. The second last one, which passes, looks like this:
```
// cdb-command: g
// cdb-command: dv
// cdb-check:[...]y = true
// cdb-check:[...]b = 0n456
// cdb-check:[...]a = 0n123
// cdb-check:[...]x = 0n42
```
and the `dv` output in the log looks like this:
```
2022-11-15T17:13:30.2292723Z 0:000>  dv
2022-11-15T17:13:30.2292998Z               a = 0n123
2022-11-15T17:13:30.2293303Z               x = 0n42
2022-11-15T17:13:30.2293590Z               b = 0n456
2022-11-15T17:13:30.2293871Z               y = true
```
But the last one, which fails, looks like this:
```
// cdb-command: g
// cdb-command: dv
// cdb-check:[...]z = 0n10
// cdb-check:[...]c = 0n789
// cdb-check:[...]y = true
// cdb-check:[...]b = 0n456
// cdb-check:[...]a = 0n123
// cdb-check:[...]x = 0n42
```
and the `dv` output in the log looks like this:
```
2022-11-15T17:13:30.2297142Z 0:000>  dv
2022-11-15T17:13:30.2297406Z               z = 0n10
2022-11-15T17:13:30.2297692Z               c = 0n789
2022-11-15T17:13:30.2297995Z               a = 0n123
2022-11-15T17:13:30.2298267Z               x = 0n42
2022-11-15T17:13:30.2298547Z               b = 0n456
2022-11-15T17:13:30.2298826Z               y = true
2022-11-15T17:13:30.2299118Z 0:000> 
2022-11-15T17:13:30.2299378Z               z = 0n10
2022-11-15T17:13:30.2299661Z               c = 0n789
2022-11-15T17:13:30.2299966Z               a = 0n123
2022-11-15T17:13:30.2300240Z               x = 0n42
2022-11-15T17:13:30.2300519Z               b = 0n456
2022-11-15T17:13:30.2300816Z               y = true
```
I don't know why the values are repeated, but maybe that's the problem? It's semi-plausible that merging basic blocks might somehow affect the debuginfo in a way that causes this...

Unfortunately I don't have access to a Windows machine which makes this hard to debug.

nnethercote commented:
@ehuss suggested the problem is the ordering of the debugger output. I have tweaked that and added a temporary commit to do more Windows testing on CI.

nnethercote commented:
Slightly reordering the debuginfo output fixed the test failure.

@bors r=bjorn3

bors commented:
:pushpin: Commit 54082dd21695373b770bd11e042e564396598c2a has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 54082dd21695373b770bd11e042e564396598c2a -->
<!-- homu: {"type":"Approved","sha":"54082dd21695373b770bd11e042e564396598c2a","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Manishearth commented:
@bors p=1

going to close the tree for non-nevers for a while

bors commented:
:hourglass: Testing commit 54082dd21695373b770bd11e042e564396598c2a with merge 251831ece9601d64172127b6caae9087358c2386...
<!-- homu: {"type":"BuildStarted","head_sha":"54082dd21695373b770bd11e042e564396598c2a","merge_sha":"251831ece9601d64172127b6caae9087358c2386"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3484689764/jobs/5830970513)
Approved by: bjorn3
Pushing 251831ece9601d64172127b6caae9087358c2386 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3484689764/jobs/5830970513"},"merge_sha":"251831ece9601d64172127b6caae9087358c2386"} -->

rust-timer commented:
Finished benchmarking commit (251831ece9601d64172127b6caae9087358c2386): [comparison URL](https://perf.rust-lang.org/compare.html?start=bebd57a9602e48431c90274fbf7d96683b0708b6&end=251831ece9601d64172127b6caae9087358c2386&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bebd57a9602e48431c90274fbf7d96683b0708b6&end=251831ece9601d64172127b6caae9087358c2386&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-1.4%, -0.4%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.2%] | 2     |
| All ❌✅ (primary)                 | -0.8% | [-1.4%, -0.4%] | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bebd57a9602e48431c90274fbf7d96683b0708b6&end=251831ece9601d64172127b6caae9087358c2386&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.6%  | [3.5%, 5.5%]   | 3     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.9%, -0.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-1.9%, -0.1%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bebd57a9602e48431c90274fbf7d96683b0708b6&end=251831ece9601d64172127b6caae9087358c2386&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
